common/src/main/java/io/netty/util/concurrent/DefaultPromise.java

```diff
@@ -454,7 +470,10 @@ protected StringBuilder toStringBuilder() {
      * @return The executor used to notify listeners when this promise is complete.
      */
     protected EventExecutor executor() {
-        return executor;
+        if (!(executor instanceof EventExecutor)) {
```

Comment:
This should be safe as people will use it from an sub-class and so should know if it is constructed via an `EventExecutor` or not.

Reply:
instanceof vs interfaces in the hot path...I can already feel the pain...:D

Reply:
haha... we could also just let it fail when casting and rethrow it... let me do this